// ng-fallimg/esm2015/lib/ng-fallimg.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgFallimgDirective } from './ng-fallimg.directive';
import { NG_FALLIMG_SOURCES } from './ng-fallimg-sources';
import * as ɵngcc0 from '@angular/core';
export class NgFallimgModule {
    /**
     * @param {?} sources All the fallback sources
     * @return {?}
     */
    static forRoot(sources) {
        return {
            ngModule: NgFallimgModule,
            providers: [
                {
                    provide: NG_FALLIMG_SOURCES,
                    useValue: sources
                }
            ]
        };
    }
}
NgFallimgModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgFallimgModule });
NgFallimgModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgFallimgModule_Factory(t) { return new (t || NgFallimgModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgFallimgModule, { declarations: function () { return [NgFallimgDirective]; }, exports: function () { return [NgFallimgDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgFallimgModule, [{
        type: NgModule,
        args: [{
                declarations: [NgFallimgDirective],
                exports: [NgFallimgDirective]
            }]
    }], null, null); })();
