// ngx-restangular/esm2015/lib/ngx-restangular-http.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ngx-restangular-http.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpBackend, HttpErrorResponse, HttpResponse } from '@angular/common/http';
import { throwError } from 'rxjs';
import { RestangularHelper } from './ngx-restangular-helper';
import { catchError, filter, map } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common/http';
export class RestangularHttp {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    createRequest(options) {
        /** @type {?} */
        const request = RestangularHelper.createRequest(options);
        return this.request(request);
    }
    /**
     * @param {?} request
     * @return {?}
     */
    request(request) {
        return this.http.handle(request)
            .pipe(filter((/**
         * @param {?} event
         * @return {?}
         */
        event => event instanceof HttpResponse)), map((/**
         * @param {?} response
         * @return {?}
         */
        (response) => {
            if (!response.ok) {
                return throwError(new HttpErrorResponse(response));
            }
            return response;
        })), map((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            response.config = { params: request };
            return response;
        })), catchError((/**
         * @param {?} err
         * @return {?}
         */
        err => {
            err.request = request;
            err.data = err.error;
            err.repeatRequest = (/**
             * @param {?=} newRequest
             * @return {?}
             */
            (newRequest) => {
                return this.request(newRequest || request);
            });
            return throwError(err);
        })));
    }
}
RestangularHttp.ɵfac = function RestangularHttp_Factory(t) { return new (t || RestangularHttp)(ɵngcc0.ɵɵinject(ɵngcc1.HttpBackend)); };
RestangularHttp.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: RestangularHttp, factory: RestangularHttp.ɵfac });
/** @nocollapse */
RestangularHttp.ctorParameters = () => [
    { type: HttpBackend }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RestangularHttp, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.HttpBackend }]; }, null); })();
if (false) {
    /** @type {?} */
    RestangularHttp.prototype.http;
}
