// ngx-restangular/esm2015/lib/ngx-restangular.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ngx-restangular.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, Optional, SkipSelf, InjectionToken } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import { RESTANGULAR, RestangularFactory } from './ngx-restangular.config';
import { Restangular } from './ngx-restangular';
import { RestangularHttp } from './ngx-restangular-http';
/** @type {?} */
import * as ɵngcc0 from '@angular/core';
export const CONFIG_OBJ = new InjectionToken('configObj');
export class RestangularModule {
    /**
     * @param {?} parentModule
     */
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('RestangularModule is already loaded. Import it in the AppModule only');
        }
    }
    /**
     * @param {?=} config1
     * @param {?=} config2
     * @return {?}
     */
    static forRoot(config1, config2) {
        return {
            ngModule: RestangularModule,
            providers: [
                { provide: CONFIG_OBJ, useValue: [config1, config2] },
                { provide: RESTANGULAR, useFactory: RestangularFactory, deps: [CONFIG_OBJ] },
            ]
        };
    }
}
RestangularModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: RestangularModule });
RestangularModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function RestangularModule_Factory(t) { return new (t || RestangularModule)(ɵngcc0.ɵɵinject(RestangularModule, 12)); }, providers: [RestangularHttp, Restangular], imports: [[HttpClientModule]] });
/** @nocollapse */
RestangularModule.ctorParameters = () => [
    { type: RestangularModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(RestangularModule, { imports: function () { return [HttpClientModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RestangularModule, [{
        type: NgModule,
        args: [{
                imports: [HttpClientModule],
                providers: [RestangularHttp, Restangular]
            }]
    }], function () { return [{ type: RestangularModule, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }]; }, null); })();
