// @awesome-cordova-plugins/safari-view-controller/__ivy_ngcc__/ngx/index.js
import { __extends } from "tslib";
import { Injectable } from '@angular/core';
import { AwesomeCordovaNativePlugin, cordova } from '@awesome-cordova-plugins/core';
import { Observable } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
var SafariViewController = /** @class */ (function (_super) {
    __extends(SafariViewController, _super);
    function SafariViewController() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SafariViewController.prototype.isAvailable = function () { return cordova(this, "isAvailable", {}, arguments); };
    SafariViewController.prototype.show = function (options) { return cordova(this, "show", { "successIndex": 1, "errorIndex": 2, "observable": true }, arguments); };
    SafariViewController.prototype.hide = function () { return cordova(this, "hide", {}, arguments); };
    SafariViewController.prototype.connectToService = function () { return cordova(this, "connectToService", {}, arguments); };
    SafariViewController.prototype.warmUp = function () { return cordova(this, "warmUp", {}, arguments); };
    SafariViewController.prototype.mayLaunchUrl = function (url) { return cordova(this, "mayLaunchUrl", {}, arguments); };
    SafariViewController.pluginName = "SafariViewController";
    SafariViewController.plugin = "cordova-plugin-safariviewcontroller";
    SafariViewController.pluginRef = "SafariViewController";
    SafariViewController.repo = "https://github.com/EddyVerbruggen/cordova-plugin-safariviewcontroller";
    SafariViewController.platforms = ["Android", "iOS"];
SafariViewController.ɵfac = function SafariViewController_Factory(t) { return ɵSafariViewController_BaseFactory(t || SafariViewController); };
SafariViewController.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: SafariViewController, factory: function (t) { return SafariViewController.ɵfac(t); } });
var ɵSafariViewController_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(SafariViewController);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SafariViewController, [{
        type: Injectable
    }], null, null); })();
    return SafariViewController;
}(AwesomeCordovaNativePlugin));
export { SafariViewController };
