// ng-fallimg/esm5/lib/ng-fallimg.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, HostListener, Inject } from '@angular/core';
import { NG_FALLIMG_SOURCES } from './ng-fallimg-sources';
/**
 * Directive for handling the default image when the main image doesn't works
 */
var NgFallimgDirective = /** @class */ (function () {
    /**
     *
     * @param el ElementRef service
     * @param fallimgSources Injector token that contains all the fallback sources
     */
    function NgFallimgDirective(el, fallimgSources) {
        this.el = el;
        this.fallimgSources = fallimgSources;
    }
    /**
     * It handles the error when the main image doesn't load
     */
    /**
     * It handles the error when the main image doesn't load
     * @private
     * @return {?}
     */
    NgFallimgDirective.prototype.errorImgHandler = /**
     * It handles the error when the main image doesn't load
     * @private
     * @return {?}
     */
    function () {
        // The chosen source
        /** @type {?} */
        var source = this.fallimgSources[this.fallimg || 'default'];
        // Verify if the key exists in all fallback sources
        if (!source) {
            throw new Error("ng-fallimg error: " + this.fallimg + " doesn't exist as a source or it is an empty source");
        }
        // it checks if the charged source is different to the actual source, this avoid the infinite bucle creation
        if (source !== this.lastSource) {
            // It saves the last source for matching when the fallback source fails and retried, that avoid an infinite bucle
            this.lastSource = source;
            ((/** @type {?} */ (this.el.nativeElement))).src = source;
        }
    };
    NgFallimgDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'img[fallimg]'
                },] }
    ];
    /** @nocollapse */
    NgFallimgDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: undefined, decorators: [{ type: Inject, args: [NG_FALLIMG_SOURCES,] }] }
    ]; };
    NgFallimgDirective.propDecorators = {
        fallimg: [{ type: Input, args: ['fallimg',] }],
        errorImgHandler: [{ type: HostListener, args: ['error',] }]
    };
    return NgFallimgDirective;
}());
export { NgFallimgDirective };
if (false) {
    /**
     * Input that receive an optional key for returning the source
     * @type {?}
     */
    NgFallimgDirective.prototype.fallimg;
    /**
     * It contains the last source setted when the main source fails
     * @type {?}
     * @private
     */
    NgFallimgDirective.prototype.lastSource;
    /**
     * @type {?}
     * @private
     */
    NgFallimgDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    NgFallimgDirective.prototype.fallimgSources;
}
