// ng-fallimg/esm5/lib/ng-fallimg.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgFallimgDirective } from './ng-fallimg.directive';
import { NG_FALLIMG_SOURCES } from './ng-fallimg-sources';
var NgFallimgModule = /** @class */ (function () {
    function NgFallimgModule() {
    }
    /**
     * @param sources All the fallback sources
     */
    /**
     * @param {?} sources All the fallback sources
     * @return {?}
     */
    NgFallimgModule.forRoot = /**
     * @param {?} sources All the fallback sources
     * @return {?}
     */
    function (sources) {
        return {
            ngModule: NgFallimgModule,
            providers: [
                {
                    provide: NG_FALLIMG_SOURCES,
                    useValue: sources
                }
            ]
        };
    };
    NgFallimgModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [NgFallimgDirective],
                    exports: [NgFallimgDirective]
                },] }
    ];
    return NgFallimgModule;
}());
export { NgFallimgModule };
